File: Assets/zfoocs/Common/PairString.cs
using System;
using System.Collections.Generic;

namespace zfoocs
{
    
    public class PairString
    {
        public string key;
        public string value;

        public static PairString ValueOf(string key, string value)
        {
            var packet = new PairString();
            packet.key = key;
            packet.value = value;
            return packet;
        }
    }


    public class PairStringRegistration : IProtocolRegistration
    {
        public short ProtocolId()
        {
            return 112;
        }

        public void Write(ByteBuffer buffer, object packet)
        {
            if (packet == null)
            {
                buffer.WriteInt(0);
                return;
            }
            PairString message = (PairString) packet;
            buffer.WriteInt(-1);
            buffer.WriteString(message.key);
            buffer.WriteString(message.value);
        }

        public object Read(ByteBuffer buffer)
        {
            int length = buffer.ReadInt();
            if (length == 0)
            {
                return null;
            }
            int beforeReadIndex = buffer.ReadOffset();
            PairString packet = new PairString();
            string result0 = buffer.ReadString();
            packet.key = result0;
            string result1 = buffer.ReadString();
            packet.value = result1;
            if (length > 0) {
                buffer.SetReadOffset(beforeReadIndex + length);
            }
            return packet;
        }
    }
}

File: Assets/zfoolua/common/PairString.lua

local PairString = {}

function PairString:new()
    local obj = {
        key = "", -- string
        value = "" -- string
    }
    setmetatable(obj, self)
    self.__index = self
    return obj
end

function PairString:protocolId()
    return 112
end

function PairString:write(buffer, packet)
    if packet == nil then
        buffer:writeInt(0)
        return
    end
    buffer:writeInt(-1)
    buffer:writeString(packet.key)
    buffer:writeString(packet.value)
end

function PairString:read(buffer)
    local length = buffer:readInt()
    if length == 0 then
        return nil
    end
    local beforeReadIndex = buffer:getReadOffset()
    local packet = PairString:new()
    local result0 = buffer:readString()
    packet.key = result0
    local result1 = buffer:readString()
    packet.value = result1
    if length > 0 then
        buffer:setReadOffset(beforeReadIndex + length)
    end
    return packet
end

return PairString
